File: src/lib/ray.rs
use crate::{Point3, Vec3};

pub struct Ray {
    orig: Point3,
    dir: Vec3,
    time: f64,
}

impl Ray {
    #[inline]
    pub fn new(origin: Point3, direction: Vec3, time: f64) -> Self {
        Self {
            orig: origin,
            dir: direction,
            time,
        }
    }

    #[inline]
    pub fn origin(&self) -> Point3 {
        self.orig
    }

    #[inline]
    pub fn direction(&self) -> Vec3 {
        self.dir
    }

    #[inline]
    pub fn time(&self) -> f64 {
        self.time
    }

    #[inline]
    pub fn at(&self, t: f64) -> Point3 {
        self.orig + t * self.dir
    }
}

File: src/ray.c
#include "ray.h"

st_vec3_t ray_at(const st_ray_t * const p_ray, double t) {
    st_vec3_t tmp = vec3_scalar_mul(&p_ray->direction, t);

    return vec3_add(&p_ray->origin, &tmp);
}
